### PR TITLE
Map XLA opaque type to a scalar TensorShape in XLAShapeToTensorShape.

### PR DESCRIPTION
**Pull request by jakeh-gc**
The `variant` type is (by convention) given a scalar shape in op definitions.
The `variant` type maps to the `opaque` type in TF2XLA.
So It appears reasonable to return a scalar `TensorShape` when an opaque typed shape is given to `XLAShapeToTensorShape`.

This allows an XLA backend to return TF variant tensors for opaque backend-specific values.

### CODE REVIEW

**Comment by sanjoy** [comment]
Hi @jakeh-gc,

> This allows an XLA backend to return TF variant tensors for opaque backend-specific values.

Can you share a bit more context on this?  Do you have something working end to end that represents TF variants as opaque values in XLA?

Thanks!

CC @jpienaar @yunxing

**Review by sanjoy — CHANGES_REQUESTED** [review]
As per the comment I left on the conversation pane.

**Comment by jakeh-gc** [comment]
Hi @sanjoy,

Thanks for taking a look!

> Can you share a bit more context on this?

Sure.

I have a custom-call. This custom-call produces some additional attributes during lowering. To minimise memory consumption and maximise throughput, we need to attach this information to the corresponding gradient instruction.

For regular attributes we would simply do the following:
```
ret = gen_my_ops.my_op(x, attr=attr)
...
@ops.RegisterGradient("MyOp")
def _my_op_grad(op, *grads):
  return [
   gen_my_ops.my_grad_op(grads[0], attr=op.get_attr("attr"))
  ]
```
Sadly, our extra attribute is only knowable during lowering of the forward custom-call instruction. For this reason, the solution I've come up with is to return the attributes in an opaque tensor. This is internally implemented as an `absl::any`.

This creates an opaque data dependency between the instructions, and means we can now use the custom-call and gradient like this:
```
ret, opaque_attr = gen_my_ops.my_op(x)
...
@ops.RegisterGradient("MyOp")
def _my_op_grad(op, *grads):
  return [
   gen_my_ops.my_grad_op(grads[0], op.output[1])
  ]
```
Where the `REGISTER_OP` has changed from having an attribute to a variant input and output.

This works as-is in most cases, except when we call `control_flow_util_v2.create_new_tf_function` on a graph that outputs an `opaque` tensor. This causes `BuildComputation` to fail in `retval.GetShape`, which calls `XLAShapeToTensorShape`.

https://github.com/tensorflow/tensorflow/blob/425b7e924b180b9979457ab7f36c78f306a5eef8/tensorflow/compiler/tf2xla/xla_compiler.cc#L220-L224

We need this information to be returned and passed to the gradient of the tf function to reach the grad op.

> Do you have something working end to end that represents TF variants as opaque values in XLA?

Sort of. We only support it inside our backend and on these computation boundaries. For example, we don't allow them on the entry computation parameters or root instruction result.

tl;dr
Being able to handle this shape conversion, allows us to support custom-calls that have attributes/data produced during lowering that can be shared with other instructions.

**Comment by gbaned** [comment]
@sanjoy  Any update on this PR? Please. Thanks!

**Comment by sanjoy** [comment]
Sorry for the delay.

The PR looks fine to me, but please add a unit test.

CC @smit-hinsu in case I'm missing something.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Requested adding a test.

**Comment by gbaned** [comment]
@jakeh-gc  Can you please check @sanjoy's comments and keep us posted ? Thanks!

**Comment by jakeh-gc** [comment]
@sanjoy @gbaned Sure, I'll pick this up again shortly.

**Comment by gbaned** [comment]
@jakeh-gc  Any update on this PR? Please. Thanks!

**Comment by jakeh-gc** [comment]
Unfortunately I haven't had time to revisit this. I'll reopen when I get the time.
Thanks.
